File: LeetCode/0804. Unique Morse Code Words/solution.cpp
/**
 * 82 / 82 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 8.3 MB 
 */
static string morse[] = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
class Solution {
public:
    int uniqueMorseRepresentations(vector<string>& words) {
        unordered_set<string> used;
        for (auto &word : words) {
            string m_word = "";
            for (auto &c : word) {
                m_word += morse[c - 'a'];
            }
            used.emplace(m_word);
        }
        return used.size();
    }
};

File: LeetCode/0804. Unique Morse Code Words/solution.py
"""
82 / 82 test cases passed.
Runtime: 40 ms
Memory Usage: 14.9 MB
"""
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
        rec = set()
        for word in words:
            m_word = ""
            for c in word:
                m_word += morse[(ord(c) - ord('a'))]
            rec.add(m_word)
        return len(rec)
 

"""
82 / 82 test cases passed.
Runtime: 32 ms
Memory Usage: 15.1 MB
"""
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
        return len(set("".join(morse[ord(c) - ord('a')] for c in word) for word in words))
